executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,546 @@
+    ) {
+        let row_offset = data.row_offset().try_into().unwrap();
+        let (data, known) = data.as_mut_slices();
+
+        let mut vars = vec![T::zero(); self.var_count];
+
+        let mut eval_stack = vec![];
+        for action in &self.actions {
+            match action {
+                InterpreterAction::AssignExpression(idx, e) => {
+                    let val = e.evaluate(&mut eval_stack, &vars[..]);
+                    vars[*idx] = val;
+                }
+                InterpreterAction::ReadCell(idx, c) => {
+                    vars[*idx] = data[index(
+                        self.first_column_id,
+                        self.column_count,
+                        row_offset,
+                        c.row_offset,
+                        c.id,
+                    )];
+                }
+                InterpreterAction::ReadParam(idx, i) => {
+                    vars[*idx] = get_param(params, *i);
+                }
+                InterpreterAction::WriteCell(idx, c) => {
+                    set(
+                        self.first_column_id,
+                        self.column_count,
+                        data,
+                        row_offset,
+                        c.row_offset,
+                        c.id,
+                        vars[*idx],
+                    );
+                }
+                InterpreterAction::WriteParam(idx, i) => {
+                    set_param(params, *i, vars[*idx]);
+                }
+                InterpreterAction::WriteKnown(c) => {
+                    set_known(
+                        self.first_column_id,
+                        self.column_count,
+                        known,
+                        row_offset,
+                        c.row_offset,
+                        c.id,
+                    );
+                }
+                InterpreterAction::MachineCall(id, arguments) => {
+                    // we know it's safe to escape the references here, but the compiler doesn't, so we use unsafe
+                    let mut args = arguments
+                        .iter()
+                        .map(|a| match a {
+                            MachineCallArgumentIdx::Unknown(idx) => {
+                                let var = &mut vars[*idx] as *mut T;
+                                LookupCell::Output(unsafe { var.as_mut().unwrap() })
+                            }
+                            MachineCallArgumentIdx::Known(idx) => {
+                                let var = &vars[*idx] as *const T;
+                                LookupCell::Input(unsafe { var.as_ref().unwrap() })
+                            }
+                        })
+                        .collect::<Vec<_>>();
+                    mutable_state.call_direct(*id, &mut args[..]).unwrap();
+                }
+                InterpreterAction::Assertion(e1, e2, expected_equal) => {
+                    let lhs_value = e1.evaluate(&mut eval_stack, &vars);
```

Comment:
Maybe assert that `eval_stack` is empty? Or create a public and a private version of `evaluate` where the public one has the assert?

Reply:
added an assert at the end of the `call` function instead